libvoikko/src/tools/voikkohyphenate.1
.TH VOIKKOHYPHENATE 1 "2010-03-06"
.SH NAME
voikkohyphenate \- test program for Voikko hyphenator
.SH SYNOPSIS
.B voikkohyphenate
.RI [ options ]
.SH DESCRIPTION
.B voikkohyphenate
is a test program for hyphenator functionality in libvoikko,
library of language tools. It reads words from stdin (one word on a line) and prints
them stdout in fully hyphenated form.
.SH OPTIONS
.TP
.BI \-s c
Use
.IR c
to separate syllables. By default \- is used. If separator is set to a character
other than \-, all original characters of the hyphenated word are printed.
.TP
.BI min_hyphenated_word_length= n
Set
.I n
as the minimun length for words to hyphenate.
.TP
.BI no_ugly_hyphenation= n
.TP
.BI hyphenate_unknown_words= n
Set the value of the specified boolean option.
.I n
can be either 0 (false) or 1 (true).
.SH SEE ALSO
.B voikkospell
for common options of different Voikko test tools.
.SH AUTHOR
voikkohyphenate and this manual page were written by \%Harri \%Pitk\[:a]nen \%(hatapitk@iki.fi).


libvoikko/ChangeLog
This ChangeLog is no longer maintained. Please see Git history for changes
after version 4.0.2.

2016-01-30 Harri Pitkänen <hatapitk@iki.fi>
	* Look harder for spelling suggestions.

2016-01-24 Harri Pitkänen <hatapitk@iki.fi>
	* Require compiler capable of supporting the C++11 standard.

2016-01-22 Harri Pitkänen <hatapitk@iki.fi>
	* Add user application data folder to dictionary search path on Windows

2015-11-01 Harri Pitkänen <hatapitk@iki.fi>
	* Enable HFST backend by default.

2015-10-31 Harri Pitkänen <hatapitk@iki.fi>
	* Disable malaga backend by default.
	* Enable VFST backend and build tools by default.
	* Add separate option --enable-experimental-vfst for experimental
	  features related to VFST format.

2015-10-28 Harri Pitkänen <hatapitk@iki.fi>
	* Use Python 3 instead of Python 2 for build tools.

2015-10-27 Harri Pitkänen <hatapitk@iki.fi>
	* Remove default value for builtin dictionary path. No directories
	  will be searched for by default unless --with-dictionary-path
	  is specified.

2015-09-13 Harri Pitkänen <hatapitk@iki.fi>
	* Make the Python module single source compatible with Python 3.
	* Drop support for Python versions 2.5 and 2.6.
	* API methods accepting path arguments now work with both Unicode
	  strings and binary strings.

2015-08-28 Harri Pitkänen <hatapitk@iki.fi>
	* Autocorrect data is read from VFST transducer in format 5
	  dictionaries.

2015-06-20 Harri Pitkänen <hatapitk@iki.fi>
	* Add spelling suggestions to simple VFST spellers by using
	  weighted VFST transducers.

2015-06-16 Harri Pitkänen <hatapitk@iki.fi>
	* Add support for weighted VFST transducers.

2015-06-08 Harri Pitkänen <hatapitk@iki.fi>
	* Initial support for BCP 47 script subtag.

2015-03-29 Harri Pitkänen <hatapitk@iki.fi>
	* Tri-licensed the HFST backend with permission from Flammie Pirinen.

2015-01-03 Harri Pitkänen <hatapitk@iki.fi>
	* Search for dictionares from $HOME/Library/Spelling/voikko on OS X.

2014-12-29 Harri Pitkänen <hatapitk@iki.fi>
	* Add support for specifying multiple default dictionary locations.

2014-08-03 Harri Pitkänen <hatapitk@iki.fi>
	* New dictionary format 5 for VFST.

2013-12-04 Harri Pitkänen <hatapitk@iki.fi>
	* Grammar checker should allow numeric references to footnotes.

2013-11-21 Harri Pitkänen <hatapitk@iki.fi>
	* Increase version numbers for 3.7.

2013-10-10 Harri Pitkänen <hatapitk@iki.fi>
	* Add API functions for listing languages for which hypenators
	  or grammar checkers are available.

2013-10-08 Harri Pitkänen <hatapitk@iki.fi>
	* Add "voikkospell -L" for listing dictionary capabilities.

2013-09-03 Harri Pitkänen <hatapitk@iki.fi>
	* Internally standard and unspecified variant are not the same.

2013-08-21 Harri Pitkänen <hatapitk@iki.fi>
	* Improve Java API for distribution through Maven Central Repository.
	* Fix test suite to work with latest Automake releases.
	* Increase version numbers for 3.6.1.

2013-07-23 Harri Pitkänen <hatapitk@iki.fi>
	* Add option accept_all_uppercase to voikkospell.
	* Improve capitalization of suggestions.

2013-07-11 Harri Pitkänen <hatapitk@iki.fi>
	* Disable voikkovfstc by default since VFST is still experimental.

2013-04-08 Harri Pitkänen <hatapitk@iki.fi>
	* Add replacement suggestion ä->e.

2013-03-17 Harri Pitkänen <hatapitk@iki.fi>
	* Private use subtag is now always returned, default is
	  "standard". This will make it easier to deal with language
	  variants in applications.

2013-03-16 Harri Pitkänen <hatapitk@iki.fi>
	* Support for ZHFST spellers is no longer experimental.

2012-08-22 Harri Pitkänen <hatapitk@iki.fi>
	* Add support for using Python 3 (and drop support for using
	  Python < 2.6) as build tool.

2012-08-20 Harri Pitkänen <hatapitk@iki.fi>
	* Add simple VFST speller implementation.

2012-08-14 Harri Pitkänen <hatapitk@iki.fi>
	* Improve word splitting suggestions near capitalized words or
	  end of sentence.

2012-08-04 Harri Pitkänen <hatapitk@iki.fi>
	* mmap may be available even if getpwuid_r is not.
	* Add option --disable-buildtools

2012-08-03 Harri Pitkänen <hatapitk@iki.fi>
	* Support building without hardcoded DICTIONARY_PATH.

2012-08-01 Harri Pitkänen <hatapitk@iki.fi>
	* Support homonyms for verb negatives.

2012-06-18 Harri Pitkänen <hatapitk@iki.fi>
	* Fix false positives from extra verb check.
	* Update version numbers for 3.5.

2012-05-14 Harri Pitkänen <hatapitk@iki.fi>
	* Check for extra main verbs.

2012-05-01 Harri Pitkänen <hatapitk@iki.fi>
	* Codepoints in Latin Extended-B block are letters.

2012-04-01 Harri Pitkänen <hatapitk@iki.fi>
	* Update project files for .Net interface to .Net 4.0 since compiler for
	  3.5 does not seem to be the default on Debian unstable anymore.

2012-03-26 Harri Pitkänen <hatapitk@iki.fi>
	* Suggest replacement e->ä

2012-03-25 Harri Pitkänen <hatapitk@iki.fi>
	* Suggest more character replacements

2012-03-24 Harri Pitkänen <hatapitk@iki.fi>
	* Suggest some two character deletions

2012-03-21 Harri Pitkänen <hatapitk@iki.fi>
	* Generate more spelling suggestions

2012-03-06 Harri Pitkänen <hatapitk@iki.fi>
	* Suggest replacement i->e

2012-03-05 Harri Pitkänen <hatapitk@iki.fi>
	* Core code and VFST backend licensed under MPL 1.1 / GPL 2+ /
	  LGPL 2.1+ tri-license.

2012-02-27 Harri Pitkänen <hatapitk@iki.fi>
	* Add new option to voikkospell: -M for morphological analysis without
	  spell checking.

2012-02-12 Harri Pitkänen <hatapitk@iki.fi>
	* Added initial VFST lookup code.

2012-02-11 Harri Pitkänen <hatapitk@iki.fi>
	* Try replacing two characters in OCR suggestion mode.

2012-01-28 Harri Pitkänen <hatapitk@iki.fi>
	* Add canonical decompositions for Russian and Komi alphabet.

2012-01-25 Sjur Moshagen <sjurnm@mac.com>
	* Add support for cyrillic alphabet

2012-01-15 Harri Pitkänen <hatapitk@iki.fi>
	* Fix build with MSVC

2012-01-14 Harri Pitkänen <hatapitk@iki.fi>
	* Fix build with GCC 4.7
	* Update version numbers for 3.4.1

2011-12-15 Harri Pitkänen <hatapitk@iki.fi>
	* Be more careful with ordinal ranges.


libvoikko/src/tools/voikkohyphenate.cpp
/* The contents of this file are subject to the Mozilla Public License Version 
 * 1.1 (the "License"); you may not use this file except in compliance with 
 * the License. You may obtain a copy of the License at 
 * http://www.mozilla.org/MPL/
 * 
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 * 
 * The Original Code is Libvoikko: Library of natural language processing tools.
 * The Initial Developer of the Original Code is Harri Pitkänen <hatapitk@iki.fi>.
 * Portions created by the Initial Developer are Copyright (C) 2006 - 2010
 * the Initial Developer. All Rights Reserved.
 * 
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *********************************************************************************/

#include "../voikko.h"
#include <iostream>
#include <cstdlib>
#include <cstdio>
#include <cstring>
#include <cwchar>

#ifdef HAVE_CONFIG_H
  #include "config.h"
#endif

using namespace std;

static const int MAX_WORD_LENGTH = 5000;

static void hyphenateWord(VoikkoHandle * handle, const wchar_t * word, size_t wlen, wchar_t separator) {
	char * result = voikkoHyphenateUcs4(handle, word);
	if (result == 0) {
		cerr << "E: hyphenation failed" << endl;
		return;
	}
	
	wchar_t * hyphenatedWord = new wchar_t[wlen * 2 + 1];
	
	const wchar_t * wordPtr = word;
	wchar_t * hyphenatedPtr = hyphenatedWord;
	char * resultPtr = result;
	size_t charsLeft = wlen;
	while (charsLeft > 0) {
		if (*resultPtr != ' ') {
			*hyphenatedPtr = separator;
			hyphenatedPtr++;
		}
		if (*resultPtr != '=' || separator != L'-') {
			*hyphenatedPtr = *wordPtr;
			hyphenatedPtr++;
		}
		resultPtr++;
		wordPtr++;
		charsLeft--;
	}
	*hyphenatedPtr = L'\0';
	wcout << hyphenatedWord << endl;
	delete[] hyphenatedWord;
	voikkoFreeCstr(result);
}

static void printHelp() {
	cout << "Usage: voikkohyphenate [OPTION]..." << endl;
	cout << "Hyphenate words read from stdin." << endl;
	cout << endl;
	cout << "For complete descriptions of available options see 'man voikkohyphenate'" << endl;
}

static void printVersion() {
	#ifdef PACKAGE_VERSION
		cout << "voikkohyphenate version " << PACKAGE_VERSION << endl;
	#endif
	cout << "libvoikko version " << voikkoGetVersion() << endl;
}

int main(int argc, char ** argv) {
	const char * path = 0;
	const char * variant = "fi";
	wchar_t separator = L'-';
	
	for (int i = 1; i < argc; i++) {
		if (strcmp(argv[i], "-p") == 0 && i + 1 < argc) {
			path = argv[++i];
		}
		else if (strcmp(argv[i], "-d") == 0 && i + 1 < argc) {
			variant = argv[++i];
		}
		else if (strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0) {
			printHelp();
			exit(0);
		}
		else if (strcmp(argv[i],  "--version") == 0) {
			printVersion();
			exit(0);
		}
	}
	const char * voikkoError;
	VoikkoHandle * handle = voikkoInit(&voikkoError, variant, path);
	if (!handle) {
		cerr << "E: Initialization of Voikko failed: " << voikkoError << endl;
		return 1;
	}
	
	voikkoSetBooleanOption(handle, VOIKKO_OPT_NO_UGLY_HYPHENATION, 0);
	
	for (int i = 1; i < argc; i++) {
		if (strcmp(argv[i], "no_ugly_hyphenation=1") == 0)
			voikkoSetBooleanOption(handle, VOIKKO_OPT_NO_UGLY_HYPHENATION, 1);
		else if (strcmp(argv[i], "no_ugly_hyphenation=0") == 0)
			voikkoSetBooleanOption(handle, VOIKKO_OPT_NO_UGLY_HYPHENATION, 0);
		else if (strcmp(argv[i], "ignore_dot=1") == 0)
			voikkoSetBooleanOption(handle, VOIKKO_OPT_IGNORE_DOT, 1);
		else if (strcmp(argv[i], "ignore_dot=0") == 0)
			voikkoSetBooleanOption(handle, VOIKKO_OPT_IGNORE_DOT, 0);
		else if (strcmp(argv[i], "hyphenate_unknown_words=1") == 0)
			voikkoSetBooleanOption(handle, VOIKKO_OPT_HYPHENATE_UNKNOWN_WORDS, 1);
		else if (strcmp(argv[i], "hyphenate_unknown_words=0") == 0)
			voikkoSetBooleanOption(handle, VOIKKO_OPT_HYPHENATE_UNKNOWN_WORDS, 0);
		else if (strncmp(argv[i], "min_hyphenated_word_length=", 27) == 0) {
			int minhwlen = atoi(argv[i] + 27);
			if (minhwlen < 2) {
				minhwlen = 2;
			}
			voikkoSetIntegerOption(handle, VOIKKO_MIN_HYPHENATED_WORD_LENGTH, minhwlen);
		}
		else if (strncmp(argv[i], "-s", 2) == 0) {
			if (strlen(argv[i]) != 3 || mbtowc(&separator, argv[i] + 2, 1) < 1) {
				cerr << "Invalid separator argument for option -s" << endl;
				return 1;
			}
		}
		else if (strcmp(argv[i], "-p") == 0 || strcmp(argv[i], "-d") == 0) {
			i++;
			continue;
		}
		else {
			cerr << "Unknown option " << argv[i] << endl;
			return 1;
		}
	}
	
	wchar_t * line = new wchar_t[MAX_WORD_LENGTH + 1];
	if (!line) {
		cerr << "E: Out of memory" << endl;
	}
	
	setlocale(LC_ALL, "");
	wcout.imbue(locale(""));
	while (fgetws(line, MAX_WORD_LENGTH, stdin)) {
		size_t lineLen = wcslen(line);
		if (lineLen == 0) {
			continue;
		}
		if (line[lineLen - 1] == L'\n') {
			line[lineLen - 1] = L'\0';
			lineLen--;
		}
		if (lineLen > LIBVOIKKO_MAX_WORD_CHARS) {
			cerr << "E: Too long word" << endl;
			continue;
		}
		hyphenateWord(handle, line, lineLen, separator);
	}
	int error = ferror(stdin);
	if (error) {
		cerr << "E: Error while reading from stdin" << endl;
	}
	delete[] line;
	
	voikkoTerminate(handle);
	return 0;
}
